packages/presets/src/ai/entries/format-bar/config.ts

```diff
@@ -154,6 +166,60 @@ const DraftAIGroup: AIItemGroupConfig = {
   ],
 };
 
+// actions that initiated from a note in edgeless mode
+// 1. when running in doc mode, call requestRunInEdgeless (let affine to show toast)
+// 2. when running in edgeless mode
+//    a. get selected in the note and let the edgeless action to handle it
+//    b. insert the result using the note shape
+function edgelessHandler<T extends keyof BlockSuitePresets.AIActions>(
+  id: T,
+  variants?: Omit<
+    Parameters<BlockSuitePresets.AIActions[T]>[0],
+    keyof BlockSuitePresets.AITextActionOptions
+  >
+) {
+  return (host: EditorHost) => {
+    if (host.doc.root?.id === undefined) return;
+    const edgeless = (
+      host.view.getWidget(
+        EDGELESS_ELEMENT_TOOLBAR_WIDGET,
+        host.doc.root.id
+      ) as EdgelessElementToolbarWidget
+    )?.edgeless;
+
+    if (!edgeless) {
+      AIProvider.slots.requestRunInEdgeless.emit({ host });
+    } else {
+      const currentController = edgeless.tools.controllers[
```

Comment:
should switch to `copilot` mode
before this line, add `edgeless.tools.setEdgelessTool({ type: 'copilot' });`